src/screens/Messages/Conversation/MessagesList.tsx

```diff
@@ -72,36 +74,63 @@ function keyExtractor(item: ConvoItem) {
   return item.key
 }
 
-function onScrollToEndFailed() {
+function onScrollToIndexFailed() {
   // Placeholder function. You have to give FlatList something or else it will error.
 }
 
 export function MessagesList() {
   const chat = useChat()
   const flatListRef = useRef<FlatList>(null)
-  // We use this to know if we should scroll after a new clop is added to the list
-  const isAtBottom = useRef(false)
-  const currentOffset = React.useRef(0)
 
-  const onContentSizeChange = useCallback(() => {
-    if (currentOffset.current <= 100) {
-      flatListRef.current?.scrollToOffset({offset: 0, animated: true})
-    }
-  }, [])
+  // We need to keep track of when the scroll offset is at the bottom of the list to know when to scroll as new items
+  // are added to the list. For example, if the user is scrolled up to 1iew older messages, we don't want to scroll to
+  // the bottom.
+  const isAtBottom = React.useRef(true)
 
-  const onEndReached = useCallback(() => {
-    if (chat.status === ConvoStatus.Ready) {
-      chat.fetchMessageHistory()
-    }
-  }, [chat])
+  // contentOpacity is a web-only hack. There is a slight delay between the rendering of the items and the user being
+  // scrolled to the bottom of the view. If we don't hide the items, there's a jarring flash whenever the scroll
+  // happens. Therefore, let's not show the items until that first scroll has actually happened
+  const [hasInitiallyScrolled, setHasInitiallyScrolled] = React.useState(false)
 
-  const onInputFocus = useCallback(() => {
-    if (!isAtBottom.current) {
-      flatListRef.current?.scrollToOffset({offset: 0, animated: true})
-    }
-  }, [])
+  // Used to keep track of the current content height. We'll need this in `onScroll` so we know when to start allowing
+  // onStartReached to fire.
+  const contentHeight = React.useRef(0)
+
+  // This is only used on native because `Keyboard` can't be imported on web. On web, an input focus will immediately
+  // trigger scrolling to the bottom. On native however, we need to wait for the keyboard to present before scrolling,
+  // which is what this hook listens for
+  useScrollToEndOnFocus(flatListRef)
```

Comment:
On web it's really simple to properly scroll to the bottom on input focus since there's no virtual keyboard. On native however, we don't want to scroll to the bottom until the keyboard has finished presenting. We use this hook to listen for the `keyboardDidShow` event, which will then trigger a scroll to bottom.

Have to put this in a hook because we cannot import `Keyboard` on web. Web is just an empty function.